### ISSUE TITLE
Infinite processing with 1.1.167 (never ends)

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#2298]
I have 4 projects that I'm currently maintaining that I run `pyright` on their base directories almost everyday to check the typing.

After updating to the latest version, one of them starts to run and never finishes. Like, never! It keeps running, without giving me any errors or logs, I can see that one CPU's core is being 100% used, but it never finishes. I waited for almost half an hour and it didn't finish... All that I can see in the output is:

```
No configuration file found.
pyproject.toml file found at /home/bellini/dev/veroo/backend.
Loading pyproject.toml file at /home/bellini/dev/veroo/backend/pyproject.toml
No include entries specified; assuming /home/bellini/dev/veroo/backend
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
stubPath /home/bellini/dev/veroo/backend/typings is not a valid directory.
Searching for source files
Auto-excluding /home/bellini/dev/veroo/backend/.venv
Found 483 source files
```

I'm also attaching the verbose output (didn't want to paste it here because it is very long), but it doesn't contain any extra information. After that output it just keeps running indefinitely .

[pyright_verbose.txt](https://github.com/microsoft/pyright/files/7157778/pyright_verbose.txt)

I don't really know how to give a better bug description here. The source code is proprietary and because of that I cannot share it here, and the other 3 projects that I mentioned are very alike (they have the same base infrastructure and have 90% of the same dependencies) and they don't present that same issue. But I'm open to do something locally to try to give you more info if there's a way.

**Comment by erictraut** [comment]
Sorry for the regression, and thanks for reporting the problem.

I think I know which change caused this. It's related to a change that I made in the code flow analyzer that's used to track types of symbols within a function. The code was previously failing to calculate some types properly when loops were involved. The new code guarantees correct types, but it can result in more computation. In certain circumstances (e.g. deeply nested loops, complex try/except statements within loops, and large numbers of if/elif/else chains within loops), the time can grow combinatorially. I suspect that's what's happening. With enough time, it will probably complete. But it's clearly a problem if it's taking minutes to complete.

I'd like to better understand the nature of the code that's triggering this problem. That's going to require us to narrow it down to a specific module and function. Since the code is proprietary, I'll need some help from you (thanks for offering!).

We have detailed trace logging in the VS Code extension (either pylance or pyright) that can help us here. Unfortunately, we don't currently expose all of this same logging from the command-line version of pyright. Do you use VS Code and have one of these two extensions loaded?

Here are the steps. Open your project in VS Code (i.e. open the top-level directory that contains the source for your project). Create a `.vscode/settings.json` file if one doesn't already exist. Add the following settings:
```json
{
  "python.analysis.diagnosticMode": "workspace",
  "python.analysis.logLevel": "Trace",
  "python.analysis.logTypeEvaluationTime": true,
}
```

This will tell pylance/pyright to type check all of the files in your workspace, similar to command-line version. It also tells it to output detailed trace logs to the Output window in VS Code (in the tab labeled either "Pyright" or "Python Language Server", depending on whether you're using pyright or pylance). And it tells it to output extremely detailed timing details for the type analyzer. You should see it spew out lots of logs until it hits the file that it's struggling with. The last line it outputs will give you a clue about which file, function, and expression it's stuck on. It would be great if you could extract that function into a semi-independent piece of code that exhibits the problem.

**Comment by erictraut** [comment]
I made a change in 1.1.168 that could potentially mitigate the perf issue you're seeing. Could you please update to the latest version and let me know what you see? Thanks!

**Comment by bellini666** [comment]
Hi @erictraut .

I'm not using vscode (I use vim and coc with https://github.com/fannheyward/coc-pyright plugin) so that config didn't do anything for me. I actually tried to see if running `pyrigh` by hand would give me extra output, but it did not. Do you know if I can put those config in pyproject.toml or pass it through cli options?

And yes, the latest release fixed the issue. It is still taking more time than before, but now it finishes in a pseudo "okishi time":

```
No configuration file found.
pyproject.toml file found at /home/bellini/dev/veroo/backend.
Loading pyproject.toml file at /home/bellini/dev/veroo/backend/pyproject.toml
No include entries specified; assuming /home/bellini/dev/veroo/backend
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
stubPath /home/bellini/dev/veroo/backend/typings is not a valid directory.
Searching for source files
Auto-excluding /home/bellini/dev/veroo/backend/.venv
Found 483 source files
0 errors, 0 warnings, 0 infos 
Completed in 125.287sec
```

(this is running in a Intel(R) Core(TM) i5-9400 CPU @ 2.90GHz and a NVMe SSD)

If there's a way to set those settings I can try to check where the code is taking longer here and try my best to extract that function into a semi-independent piece of code.

Thanks so far!

**Comment by jakebailey** [comment]
The settings above are just settings that go over LSP; your editor or its plugins should provide a way to set configurations. Skimming `coc-pyright`'s repo, their README says that you can do this in `coc-settings.json`.

**Comment by bellini666** [comment]
@jakebailey yeah, but when running `pyright` from the cli it will not use the editor's config, because in the editor I don't have easy access to the LSP's logs.

**Comment by erictraut** [comment]
There's current no way to enable these verbose logs through the CLI. If you follow the instructions above, you should be able to replicate it in the editor. If you can't make it work in vim, perhaps you'd be willing to install VS Code for the purposes of diagnosing this issue? It's a free download and install.

**Comment by mikhail-mmlc** [comment]
I'm having a similar issue with one of python files. The time went up for this particular file from `10.505sec` in `pyright 1.1.166` to `11065.224sec` in `pyright 1.1.168`. 
I'm using the CLI but I could build it with extra log output if needed.

**Comment by bellini666** [comment]
@erictraut I see. Sure, I can download vscode and try to get that extra log output. I'm not used to it though, can you explain to me how to install a specific pyright/pylance version in it? Also, where will I be able to get that log output?

**Comment by jakebailey** [comment]
Pylance hasn't been updated since the changes here would made, so you can just install Pyright directly at its latest version. You'll find logs in the "pyright" output window (which you can open up by hitting `Ctrl+Shift+P` to open the command palette and searching for "output", then selecting it at the dropdown at the right).

**Comment by jakebailey** [comment]
@mikhail-mmlc Do you have this file somewhere we could access it to test? We don't have a repro for this yet, so that would be very helpful.

**Comment by bellini666** [comment]
Here are the logs, but it doesn't seem like pylance took too much to generate those, it was about 15 seconds until it became idle. Note that pyright 1.1.167 was taking more than 30 minute (it probably would finish some time, but I didn't wait) and 1.1.168 is taking 125 seconds, so I'm not sure that this pylance version has the bug. But either way, here is the log results:

[logs.txt](https://github.com/microsoft/pyright/files/7165745/logs.txt)

If there's a way to install a specific pylance version in vscode I can do that for you and get better logs.

**Comment by erictraut** [comment]
Pylance won't exhibit the problem because it hasn't been published with this change. Could you please uninstall pylance and install the pyright VS Code extension instead?

**Comment by mikhail-mmlc** [comment]
> @mikhail-mmlc Do you have this file somewhere we could access it to test? We don't have a repro for this yet, so that would be very helpful.

I'm afraid I can't share the file with you. However, I could put debug prints into the pyright source code and lint the file with the custom build. I've narrowed the issue to one big function with many nested `for`, `if` and `try` statements.
Let me know what would be a good place(s) to put debug prints.

**Comment by bellini666** [comment]
@erictraut I did it and I can see that pyright is indeed taking a lot more time to finish now (I can see it using my cpu for a long time), but now, different from pylance, I can't see anything when checking the output of the `Python Language Server` log (it is empty)

Do I need to do something different to get this?

**Comment by jakebailey** [comment]
It's under "pyright".

**Comment by erictraut** [comment]
@mikhail-mmlc, I really need a repro case. If you know which function is causing the problem, is there a chance you could send me a sanitized copy of that function?

**Comment by bellini666** [comment]
Ah ok, found it now. Here are the logs:

[pyright_output.txt](https://github.com/microsoft/pyright/files/7170492/pyright_output.txt)

It is using the 1.1.168 version so it took a while (~5 minutes) to complete, but not indefinitely like 1.1.167 would take

**Comment by erictraut** [comment]
@belllini666, thanks for the logs. They indicate that the problem is with a file called `report.py` and a method called `write_pdf`. This method appears to be very complex, with many unannotated variables, nested loops, etc. Does that sound correct? In particular, there's an inner loop that is assigning a value to a variable called `lost_space` that seems to be contributing to the long analysis times.

Would it be possible for you to post a (sanitized) copy of that method? I really need a repro case to diagnose the problem. If it's not feasible to post that method, the inner loop (with the `lost_space` assignment) might give me some clues, but I don't know if it will be enough to fully diagnose.

**Comment by bellini666** [comment]
Hey @erictraut , that method is not actually inside there, it is from weasyprint, a library that we use to generate PDFs from html+css files. Here is the snipped where it is used in my code inside that `report.py` file:

```python
    html = weasyprint.HTML(
        string=render_to_string(
            os.path.join("reports", f"delivery_report_{kind}.html"),
            {
                "date": timezone.localtime().strftime("%d/%m/%Y %H:%M:%S"),
                "kind": kind,
                "volumes_total": volumes_total,
                "orders": orders_data,
            },
        ),
    )
    ret = html.write_pdf(target=None)
```

That `render_to_string` uses django templating language to generate an html file (similar to jinja2) as a string, which is passed to `weasyprint.HTML` and then we call that method to generate a PDF from it (since the `target` is `None`, it will be rendered in memory)

You can probably check the weasyprint code (and that method in specific) in here: https://github.com/Kozea/WeasyPrint/blob/master/weasyprint/__init__.py#L138

Please let me know if you need anything else from me, I'll gladly provide more info it I can.

**Comment by erictraut** [comment]
Great! I'm able to repro the problem. That should allow me to investigate further. Thanks for your help!

This library is untyped. I presume that you've enabled `useLibraryCodeForTypes`, telling pyright that you want it to try to infer types from the library's source code. We typically recommend against enabling this if you are using type checking because inferred types are often not accurate enough and lead to false positive type errors. I guess that it must have been working well enough for you (prior to this performance regression)?

**Comment by mikhail-mmlc** [comment]
> @mikhail-mmlc, I really need a repro case. If you know which function is causing the problem, is there a chance you could send me a sanitized copy of that function?

[Done](https://gist.github.com/mikhail-mmlc/efd7e83cd27548aa8308ab2bf4809c35).
pyright version:  `1.1.167`
total time: `173.435sec`, but it takes just `4s` if you comment all `logger.<name>` statements.

**Comment by erictraut** [comment]
Thanks @mikhail-mmlc, I'm able to repro the problem with your example as well. Very helpful!

**Comment by bellini666** [comment]
@erictraut , yes, I'm indeed using `useLibraryCodeForTypes`. I'll think about disabling it if it doesn't have any bad side effects, but yes, it was working well enough before (like, not blazing fast, far from it, but it was acceptable)

**Comment by erictraut** [comment]
I've implemented several optimizations that eliminate the performance regression. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.169. It will also be included in the next release of pylance.

**Comment by bellini666** [comment]
Yes, indeed much better!

```
No configuration file found.
pyproject.toml file found at /home/bellini/dev/veroo/backend.
Loading pyproject.toml file at /home/bellini/dev/veroo/backend/pyproject.toml
No include entries specified; assuming /home/bellini/dev/veroo/backend
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
stubPath /home/bellini/dev/veroo/backend/typings is not a valid directory.
Searching for source files
Auto-excluding /home/bellini/dev/veroo/backend/.venv
Found 485 source files
0 errors, 0 warnings, 0 infos 
Completed in 22.167sec
```

Thanks @erictraut , will let you know if I hit any other issues!
